File: services/python/edtech_service/stores/meetings_store.py
from __future__ import annotations
from dataclasses import dataclass, field
from typing import List, Optional, Dict, Any
import hashlib


@dataclass
class MeetingRecord:
    instructor_id: str
    course_id: str
    start_time: str
    participants: List[Dict[str, str]] = field(default_factory=list)
    uid: str = ""
    sequence: int = 1


class MeetingsStore:
    def __init__(self) -> None:
        self._by_id: Dict[str, MeetingRecord] = {}

    def _default_uid(self, meeting_id: str) -> str:
        h = hashlib.sha1(meeting_id.encode("utf-8")).hexdigest()
        return f"{h}@edtech.py.local"

    def upsert(
        self,
        meeting_id: str,
        *,
        instructor_id: Optional[str] = None,
        course_id: Optional[str] = None,
        start_time: Optional[str] = None,
        participants: Optional[List[Dict[str, str]]] = None,
    ) -> MeetingRecord:
        rec = self._by_id.get(meeting_id)
        if rec:
            if instructor_id is not None:
                rec.instructor_id = instructor_id
            if course_id is not None:
                rec.course_id = course_id
            if start_time is not None:
                rec.start_time = start_time
            if participants is not None:
                rec.participants = participants
            self._by_id[meeting_id] = rec
            return rec
        rec = MeetingRecord(
            instructor_id=instructor_id or "",
            course_id=course_id or "",
            start_time=start_time or "",
            participants=participants or [],
            uid=self._default_uid(meeting_id),
            sequence=1,
        )
        self._by_id[meeting_id] = rec
        return rec

    def get(self, meeting_id: str) -> Optional[MeetingRecord]:
        return self._by_id.get(meeting_id)

    def bump_sequence(self, meeting_id: str) -> int:
        rec = self._by_id.get(meeting_id)
        if not rec:
            return 0
        rec.sequence = (rec.sequence or 1) + 1
        self._by_id[meeting_id] = rec
        return rec.sequence

    def clear(self) -> None:
        self._by_id.clear()


store = MeetingsStore()

File: services/node/src/stores/meetingsStore.js
// Simple in-memory meeting store. Replace with DB in production.
const crypto = require('crypto');

class MeetingsStore {
  constructor() {
    this.byId = new Map(); // meetingId -> { instructor_id, course_id, start_time, participants, uid, sequence }
  }

  _defaultUID(meetingId) {
    const hash = crypto.createHash('sha1').update(String(meetingId)).digest('hex');
    return `${hash}@edtech.local`;
  }

  upsert(meetingId, { instructor_id, course_id, start_time, participants }) {
    const existing = this.byId.get(meetingId);
    if (existing) {
      // Keep existing UID/sequence, update fields
      existing.instructor_id = instructor_id ?? existing.instructor_id;
      existing.course_id = course_id ?? existing.course_id;
      existing.start_time = start_time ?? existing.start_time;
      existing.participants = Array.isArray(participants) ? participants : existing.participants;
      this.byId.set(meetingId, existing);
      return existing;
    }
    const record = {
      instructor_id,
      course_id,
      start_time,
      participants: Array.isArray(participants) ? participants : [],
      uid: this._defaultUID(meetingId),
      sequence: 1,
    };
    this.byId.set(meetingId, record);
    return record;
  }

  get(meetingId) {
    return this.byId.get(meetingId) || null;
  }

  bumpSequence(meetingId) {
    const r = this.byId.get(meetingId);
    if (!r) return null;
    r.sequence = (r.sequence || 1) + 1;
    this.byId.set(meetingId, r);
    return r.sequence;
  }
}

module.exports = new MeetingsStore();
